### PR TITLE
add Close and DrainContext

### PR DESCRIPTION
**Pull request by jackbearheart**
Don't use a sync.WaitGroup because sync.WaitGroup will panic
if wg.Add is called while another goroutine is in wg.Done().
In our real life programs, we often accidentally (or intentionally)
leave goroutines running that may log, but I still will want to
"flush as much as I can" before exiting the program. This change
gives you some options for shutdown.

### CODE REVIEW

File: logger/logger_writer.go [hunk 1]
```diff
@@ -225,26 +223,61 @@ func logWriter() {
 			writeCustomSocket(msg)
 		}
 		freeMsg(msg)
-
-		pendingRecordWG.Done()
 	}
 	if customSock != nil {
 		customSock.Close()
 	}
+	logWriterFinished <- struct{}{}
```
Comment (theclapp) [inline logger/logger_writer.go]: Why not `close(logWriterFinished)`?

File: logger/logger_writer.go [hunk 2]
```diff
@@ -225,26 +223,61 @@ func logWriter() {
...
+		select {
+		case <-ctx.Done():
+			return
+		default:
+		}
+		time.Sleep(10 * time.Millisecond)
+		if len(messages) <= 0 {
+			return
+		}
+	}
```
Comment (theclapp) [inline logger/logger_writer.go]:
```go
select {
case <-ctx.Done():
	return
default:
}
```

This is equivalent to `if ctx.Err() != nil { return }`, yes?

As for the sleep -- why not create a new context, based on the given one, with a 10ms timeout?

```go
for ctx.Err() == nil && len(messages) > 0 {
	innerCtx, cancel := context.WithTimeout(ctx, 10*time.Millisecond)
	<-innerCtx.Done()
	cancel()
}
```

File: logger/logger_writer.go (hunk 1 above)
Comment (jackbearheart) [reply logger/logger_writer.go]: Oops. You're right!

File: logger/logger_writer.go (hunk 2 above)
Comment (jackbearheart) [reply logger/logger_writer.go]: That's smart too. Will do.

File: logger/logger_writer.go [hunk 3]
```diff
@@ -227,40 +227,37 @@ func logWriter() {
...
+		cancel()
 	}
+	return ctx.Err()
 }
 
 // Drain is like DrainContext, but you didn't want to write context.Background().
 // Outside of tests, you want to use DrainContext.
 func Drain() {
-	DrainContext(context.Background())
+	_ = DrainContext(context.Background())
```
Comment (theclapp) [inline logger/logger_writer.go]: Are you just trying to signal to the reader that you're ignoring the error on purpose?

Comment (theclapp) [reply logger/logger_writer.go]: I mean, if so, carry on.  Just curious.

Comment (jackbearheart) [reply logger/logger_writer.go]: Yes, that's it.

**Comment by jackbearheart** [comment]
@i3149 seem alright?

**Comment by jackbearheart** [comment]
This got ignored and wasn't high priority for me. I'm gonna take lack of interest and me having tested this as an excuse to merge this week. Objections? Speak now or forever hold your peace.

**Comment by i3149** [comment]
Merge away.  :)

**Pull request merged**